### PR TITLE
Implement ViT, several variants of ViT, ViT for double images input and several miscellaneous

### PR DESCRIPTION
**Pull request by ShuoyuChen-physics**
1. cnn_dataset.py 
    I updated some optimization method, and you can check the detail in the comment.
2. Relative error  loss function
    A better loss function uses for momentum regression.
3. ViT models
    I added four ViT models in watchmal/model.
    vitransformer.py: traditional ViT
    SwinTransformer.py: Shifted Window Transformer (here I used 'timm' library, but it's not in the apptainer.)
    T2TViTransformer.py: Token-to-token Transformer (single image) 
    T2TViTransformerDI.py: Token-to-token Transformer (double images) (This model is designed to use  hit information from both 20inch PMTs and mPMTs. Now I'm testing this)
4. AdamW and coslineLR
    The standard optimizer and scheduler use for ViT training.

### CODE REVIEW

File: watchmal/engine/reconstruction.py [hunk 1]
```diff
@@ -81,6 +82,10 @@ def configure_loss(self, loss_config):
     def configure_optimizers(self, optimizer_config):
         """Instantiate an optimizer from a hydra config."""
         self.optimizer = instantiate(optimizer_config, params=self.module.parameters())
+        total_params = sum(p.numel() for p in self.module.parameters() if p.requires_grad)
+        opt_params = sum(p.numel() for g in self.optimizer.param_groups for p in g['params'])
+        print(f"Total trainable parameters: {total_params}")
+        print(f"Parameters passed to optimizer: {opt_params}")
```
Comment (nickwp) [inline watchmal/engine/reconstruction.py:88]: This is a nice addition, thank you!

**Comment by ShuoyuChen-physics** [comment]
I fully understand, I will do these recently.

> Alternatively, we could include the implementation of SwinT directly in WatchMaL.
I'm planning to do this, because I want to modify it to dual images input to integrate mPMT, so I have to implemeny SwinT directly in watchmal.

But I still think we need timm, because it's convenient for any new model. Even though now I don't have thought for other models, this package can work for anyone who want to try a new model quickly.

File: watchmal/loss/relative_huber.py
[the remark this answers is not in the document]
Comment (nickwp) [reply watchmal/loss/relative_huber.py:1]: Thanks, this makes sense to me and it's great to see it working after the changes.
If it's not too much trouble to train direct comparisons, it would be nice to see a direct comparison showing the effect of each individual change.

**Comment by ShuoyuChen-physics** [comment]
Hi, sorry for this late re-request. Here I've done some changes:
1. Clean up configs for ViT.
2. Use class inheritance to rewrite dual image dataset and engine.
I've not changed the CNNDataset here to use the list of channels approach. I agree that might be better, but I think it's not urgent to change and now it works well for me. I will do it later.
I appreciate your further suggestion.

**Comment by ShuoyuChen-physics** [comment]
> Thanks for these updates @ShuoyuChen-physics!
> 
> While I think some of the config files could still be cleaned up a bit (there's at least a couple of files that have a block of lines identical to eachother) we can clean it up later as part of a general config cleanup.
> 
> Now I just want to find time to add timm to the container and make sure nothing is broken, then can merge.

Thank you for your review, I just cleaned up some config files in ViT part.

**Comment by nickwp** [comment]
Maybe it's best to unmerge #85 so this can be approved and merged, then I'll update #85 accordingly before merging

**Comment by pdeperio** [comment]
but conflicts didn't seem severe?

**Comment by nickwp** [comment]
The changes needed are actually not simple, because the forward function was removed entirely and replaced by forward_pass, resulting in a substantially different structure. So I think fixing it will require a bit of work. I'm not sure which way round is easier now. Maybe easier at this point to just fix it in this PR, but still might be easier to revert the merge of #85, then merge this PR once testing is finished, then make a new PR that does the same as #85 but with the changes needed after this PR... Either way it's a bit of a mess I think.

**Comment by nickwp** [comment]
The simplest solution to the merge was actually to refactor things so that there's no need for a new DI regression engine subclass at all, so no merge conflicts in the regression engine:
I changed it so that the DI dataset provides a single `"data"` entry that is a tuple of Shuoyu's prevous `"data_main"` and `"data_second"` entries, and changed the DI model's forward method to expect a single argument, which gets unpacked into the `x_main` and `x_second` that prevously came from Shuoyu's `"data_main"` and `"data_second"` entries.
So now it can just use the usual regression engine that takes the `"data"` entry from the dictionary provided by the dataset and passes it to the model.

There's no merge conflicts now, so I'll test a locally merged version can run the tutorial ok, which should at least confirm nothing is broken.

@ShuoyuChen-physics please can you also pull and test this version after my modifications? Particularly check that using the DI dataset and DI model still works. You might need to update the config yaml to no longer use the DI regression engine (that no longer exists).

**Comment by ShuoyuChen-physics** [comment]
> The simplest solution to the merge was actually to refactor things so that there's no need for a new DI regression engine subclass at all, so no merge conflicts in the regression engine: I changed it so that the DI dataset provides a single `"data"` entry that is a tuple of Shuoyu's prevous `"data_main"` and `"data_second"` entries, and changed the DI model's forward method to expect a single argument, which gets unpacked into the `x_main` and `x_second` that prevously came from Shuoyu's `"data_main"` and `"data_second"` entries. So now it can just use the usual regression engine that takes the `"data"` entry from the dictionary provided by the dataset and passes it to the model.
> 
> There's no merge conflicts now, so I'll test a locally merged version can run the tutorial ok, which should at least confirm nothing is broken.
> 
> @ShuoyuChen-physics please can you also pull and test this version after my modifications? Particularly check that using the DI dataset and DI model still works. You might need to update the config yaml to no longer use the DI regression engine (that no longer exists).

Sorry for this late replay and thank you very much for your changes. Recently, I'm working on writing a SwinT model with DI and encounter lots of problems. So I have to do lots of changes to DI engine, DI dataset and new model. 

I think the DI part of this PR should be abandoned and now I'm testing the new DI engine, dataset and model. If there is no conflict, I don't know whether we need to pass this PR, but anyway I will make the new code compatible with the newest version and submit the new DI part after my testing, maybe after HKCM.
